docs/source/csdl-spec.md

```diff
@@ -0,0 +1,630 @@
+---
+title: CSDL Specification
+sidebar_title: CSDL Specification
+description: The Composed Schema Definition Language
+status: Draft
+version: "1.0.0"
+authors: 
+  - Jake Dawkins <jake@apollographql.com>
+  - Trevor Scheer <trevor@apollographql.com>
+  - Ashi Krishnan <ashi@apollographql.com>
+---
+
+## Introduction: What is CSDL?
+
+The Composed Schema Definition Language (CSDL) provides a way to describe a GraphQL Schema which is composed from one or more other GraphQL schemas.
+
+The CSDL is intended to be a single artifact which can power a graph router. For example, the CSDL replaces the service list configuration of Apollo Gateway.
+
+CSDL is a subset of the [GraphQL Schema Definition Language](https://spec.graphql.org/). It removes certain SDL features (for instance, type extensions are not permitted) while also requiring the definition of certain types and providing a suite of directives to describe graph topology. In particular, a CSDL can:
+- define [subgraphs](#subgraph) and bind them to [endpoints](#endpoint) (with the required [`csdl_Graph`](#csdl_Graph) type and [`@csdl_endpoint`](#csdl_endpoint) directive)
+- assign fields to subgraphs (with [`@csdl_resolve`](#resolve))
+- declare additional data required and provided by subgraph field resolvers (with [`@csdl_key`](#key), and [`@csdl_resolve`](#csdl_resolve))
+
+## How to read this document
+
+TK, discussion of spec nomenclature
+e.g.: https://html.spec.whatwg.org/multipage/introduction.html#how-to-read-this-specification
+
+### What this document isn't
+
+This document specifies the CSDL and only the CSDL. It does not have an opinion about how CSDL should be generated from subgraphs. A suggestion is provided in [Appendix: Basic Composition Algorithm](#appendix-suggested-composition-algorithm), but conforming implementations may choose approach they like.
+
+## Example: Let's go to the moon
+
+*This section is non-normative.*
+
+We'll refer to this example throughout the document. The example consists of two example schemas—a `rockets` schema which serves data about rockets, and an `astronauts` service which services queries regarding the people flying on them—along with a CSDL that composes them.
+
+### Rockets
+
+<a name=ex-rockets href=#ex-rockets class='listing not-csdl'>Rockets example service</a>
+```graphql
+# Rockets Service
+type Query {
+  rockets: [Rocket]!
+}
+
+type Rocket @key(fields: "id") {
+  id: String! 
+  name: String!
+  captain: Astronaut @provides(fields: "tripId")
```

Comment:
Assuming we're using the `@provides` to illustrate the composite nature of the (currently named) `@csdl_key` directive — should we include an `@requires` as well? I know that the overwhelming majority of _our_ federation usage relies on `@requires` and we don't have a single usage of `@provides` (our priorities are not query planner performance tricks), so I think if we're going to include things beyond the essential, `@requires` would probably be more universally understood and informative, if other teams are anything like ours.